Shear validation test: high-z bug fixes and mag cut implemented

patricialarsen opened the pull request:
The updates for the test are as follows:

- catch for missing cosmology 
- bug fix for high redshift theory
- magnitude cut to speed up test 

There was a bug stopping the theory correlations being computed above redshift 2, I've also added a magnitude cut for a magnitude quantity and limit defined in the yaml file (note that current default is for umachine set-up) to let the test run in a more reasonable time-frame. 

The portal link should be: https://portal.nersc.gov/project/lsst/descqa/v2/?run=2018-08-24_15

patricialarsen commented:
![summary](https://user-images.githubusercontent.com/20701830/44608803-cd9a5780-a7ba-11e8-972d-8eaa5a5a47a8.png)
(as portal is currently down)

yymao commented on descqa/shear_test.py:
```diff
@@ -76,11 +80,11 @@ def __init__(self,
     def compute_nz(self, n_z):
         '''create interpolated n(z) distribution'''
-        z_bins = np.linspace(0.0, 2.0, 101)
+        z_bins = np.linspace(self.zlo,self.zhi, 301)
```
add a space after comma

yymao commented on descqa/shear_test.py:
```diff
@@ -254,29 +258,38 @@ def run_on_single_catalog(self, catalog_instance, catalog_name, output_dir):
         '''
         run test on a single catalog
         '''
+        f1 = open('./logfile','w+')
```
New file must be created within `output_dir`.

yymao commented on descqa/shear_test.py:
```diff
@@ -32,6 +32,8 @@ def __init__(self,
                  dec='dec',
                  e1='shear_1',
                  e2='shear_2_phosim',
+                 mag='restframe_extincted_sdss_abs_magr',
```
is `mag` supposed to be the column name of the magnitude? If so, the default value should set to something in the schema, rather than a catalog-specific native quantitiy name

patricialarsen replied:
this is now set to Mag_true_r_sdss_z0

yymao commented on descqa/shear_test.py:
```diff
@@ -254,29 +258,38 @@ def run_on_single_catalog(self, catalog_instance, catalog_name, output_dir):
         catalog_data = self.get_catalog_data(
-            catalog_instance, [self.z, self.ra, self.dec, self.e1, self.e2, self.kappa], filters=self.filters)
-
+            catalog_instance, [self.z, self.ra, self.dec, self.e1, self.e2, self.kappa,self.mag], filters=self.filters)
```
add a space after comma

yymao commented on descqa/shear_test.py:
```diff
@@ -254,29 +258,38 @@ def run_on_single_catalog(self, catalog_instance, catalog_name, output_dir):
-            catalog_instance, [self.z, self.ra, self.dec, self.e1, self.e2, self.kappa], filters=self.filters)
-
+            catalog_instance, [self.z, self.ra, self.dec, self.e1, self.e2, self.kappa,self.mag], filters=self.filters)
+         
```
remove space

yymao commented on descqa/shear_test.py:
```diff
@@ -254,29 +258,38 @@ def run_on_single_catalog(self, catalog_instance, catalog_name, output_dir):
         #TODO: set sigma8 value to catalog value when this becomes possible
 
+        f1.write("set cosmology\n")
         pars.InitPower.set_params(ns=0.963, As=2.168e-9)
```
n_s and sigma_8 are now available in `catalog_instance.cosmology`

yymao commented on descqa/shear_test.py:
```diff
@@ -313,13 +327,16 @@ def run_on_single_catalog(self, catalog_instance, catalog_name, output_dir):
             gg.process(cat_s)
-
+            #NOTE commented out for now - will need to change back
+            #r = np.exp(gg.meanlogr)
```
remove obsolete comments

evevkovacs commented:
@cwwalter  @patricialarsen Please see the plot above in answer to you question about the level of agreement between the theory and catalog over the full range of redshifts provided in the catalog. (The plot we showed this morning was only for z=0.6) Thanks Patricia for updating this test.

yymao commented on descqa/shear_test.py:
```diff
@@ -292,17 +305,18 @@ def run_on_single_catalog(self, catalog_instance, catalog_name, output_dir):
             zhi2 = z_mean + self.z_range
             print(zlo2, zhi2)
-            zmask = (catalog_data[self.z] < zhi2)*(catalog_data[self.z] > zlo2)
+            zmask = (catalog_data[self.z]< zhi2)*(catalog_data[self.z] > zlo2)
```
use ` & ` instead of `*` to join boolean masks

yymao replied:
Also, you can define the total mask here:
```python
mask = zmask & mask_mag
```
to be used in Lines 316-319

yymao commented on descqa/shear_test.py:
```diff
@@ -292,17 +305,18 @@ def run_on_single_catalog(self, catalog_instance, catalog_name, output_dir):
+                ra=catalog_data[self.ra][zmask*mask_mag],
+                dec=catalog_data[self.dec][zmask*mask_mag],
+                g1=catalog_data[self.e1][zmask*mask_mag] - np.mean(catalog_data[self.e1][zmask*mask_mag]),
+                g2=-(catalog_data[self.e2][zmask*mask_mag] - np.mean(catalog_data[self.e2][zmask*mask_mag])),
```
use ` & ` instead of `*` to join boolean masks. Or better, save `zmask & mask_mag` in a variable in advance.

yymao commented on descqa/shear_test.py:
```diff
@@ -313,13 +327,16 @@ def run_on_single_catalog(self, catalog_instance, catalog_name, output_dir):
             xim = gg.xim * 1.e6
-
+            #xip = np.zeros((self.nbins))
+            #xim = np.zeros((self.nbins))
```
Remove unused code, or explain why they should be kept

yymao commented on descqa/shear_test.py:
```diff
@@ -342,9 +359,9 @@ def run_on_single_catalog(self, catalog_instance, catalog_name, output_dir):
                     sigm_jack[i] = np.sqrt(gg.varxi[i])*1.e6
 
-            n_z = catalog_data[self.z][zmask]
+            n_z = catalog_data[self.z][zmask*mask_mag]
```
use ` & ` instead of `*` to join boolean masks. or use pre-defined mask as discussed above.

yymao commented on descqa/shear_test.py:
```diff
@@ -357,6 +374,7 @@ def run_on_single_catalog(self, catalog_instance, catalog_name, output_dir):
             print(theory_minus)
             print(xip)
             print(xim)
+            f1.write("all done! \n")
```
need to close `f1`.

patricialarsen commented:
@yymao  thanks for the comments, I've addressed these all. f1 was just for my own debugging - this has been removed. I don't see an ns or sigma8 value in the cosmology instance - how do I access these?

yymao commented:
@patricialarsen  Here's an example of accessing `n_s` and `sigma8` of cosmoDC2
```python
cat = GCRCatalogs.load_catalog('cosmoDC2_v0.1')
cat.cosmology.n_s
cat.cosmology.sigma8
```

If you didn't find `n_s` and `sigma8` in the catalog that you are using, you can open an issue in https://github.com/LSSTDESC/gcr-catalogs

yymao commented:
@patricialarsen also, I think the adding the native quantity mag name in the test config is also not ideal. We should really be using the mag name in the schema.

yymao commented:
Lines 185-188: Store the mask `labs != i` for repeated use.

yymao commented:
@patricialarsen sorry for my unorganized review comments. Most of them came to me after I reading a bit more into the code. 

You mentioned about the redshift range won't work for protoDC2. Is that mostly due to how the tomographic bins are set up (i.e., https://github.com/LSSTDESC/descqa/blob/22b4001/descqa/shear_test.py#L303)? Can we check the max redshift (`:= z_max`) in the catalog, and if less then `self.zhi`, replace `self.zhi` with `z_max`?

patricialarsen commented:
@yymao sorry for the delay in this - in the latest commit I fixed the maximum redshift as suggested and think I fixed all the other minor issues you mentioned but that was a week or two back so I should probably recheck and make sure I didn't miss any.

patricialarsen commented:
@yymao I have double-checked the corrections mentioned in your review, updated with respect to the master, and verified that this still runs. The travis build doesn't complete seemingly due to an issue with the camb installation. To remind you of the changes here:

- maximum redshift gets corrected to maximum survey redshift to prevent crash on prototype catalog

- cosmological parameters and unit corrections updated

- magnitude cuts implemented (to reduce compute time)

- catch for missing cosmology

There's one new change, which is to incorporate the redshift and magnitude cuts properly into the jackknife part of the code so that it's consistent with the other updates. At least on my end I think this pull request is ready to be merged.

yymao commented on descqa/shear_test.py:
```diff
@@ -12,11 +12,10 @@
 import astropy.units as u
```
```suggestion
import astropy.units as u # pylint: disable=no-member
```

yymao commented on descqa/shear_test.py:
```diff
@@ -12,11 +12,10 @@
 import astropy.units as u
 import astropy.constants as const
```
```suggestion
import astropy.constants as const # pylint: disable=no-member
```

yymao commented on descqa/shear_test.py:
```diff
@@ -100,42 +101,41 @@ def galaxy_W(self, z, n, chi_int, cosmo, chi_recomb):
         prefactor = cst * chi * (1. + z) * u.Mpc
         val_array = []
         for i in range(len(z)):
```
```suggestion
        for chi_this in chi:
```

yymao commented on descqa/shear_test.py:
```diff
@@ -100,42 +101,41 @@ def galaxy_W(self, z, n, chi_int, cosmo, chi_recomb):
         val_array = []
         for i in range(len(z)):
-            val_array.append(quad(self.integrand_w, chi[i], chi_recomb, args=(n, chi[i], chi_int, cosmo))[0])
+            val_array.append(quad(self.integrand_w, chi[i], chi_max, args=(n, chi[i], chi_int, cosmo))[0])
```
```suggestion
            val_array.append(quad(self.integrand_w, chi_this, chi_max, args=(n, chi_this, chi_int, cosmo))[0])
```

yymao commented on descqa/shear_test.py:
```diff
@@ -259,24 +259,30 @@ def run_on_single_catalog(self, catalog_instance, catalog_name, output_dir):
-        pars.InitPower.set_params(ns=0.963, As=2.168e-9)
-        #pars.InitPower.set_params(ns=0.963,As = 2.168e-9*(sigma8/0.8 )**2)
+        pars.set_cosmology(H0=cosmo.H0.value, ombh2=cosmo.Ob0*cosmo.h**2, omch2=(cosmo.Om0-cosmo.Ob0)*cosmo.h**2)
+        pars.InitPower.set_params(ns=ns, As = 2.168e-9*(s8/0.8)**2)
```
```suggestion
        pars.InitPower.set_params(ns=ns, As=2.168e-9*(s8/0.8)**2)
```

yymao commented on descqa/shear_test.py:
```diff
@@ -259,24 +259,30 @@ def run_on_single_catalog(self, catalog_instance, catalog_name, output_dir):
-        #pars.InitPower.set_params(ns=0.963,As = 2.168e-9*(sigma8/0.8 )**2)
+        pars.set_cosmology(H0=cosmo.H0.value, ombh2=cosmo.Ob0*cosmo.h**2, omch2=(cosmo.Om0-cosmo.Ob0)*cosmo.h**2)
+        pars.InitPower.set_params(ns=ns, As = 2.168e-9*(s8/0.8)**2)
         camb.set_halofit_version(version='takahashi')
```
```suggestion
        camb.set_halofit_version(version='takahashi') # pylint: disable=no-member
```

yymao commented on descqa/shear_test.py:
```diff
@@ -76,22 +77,22 @@ def __init__(self,
         n2 = interp1d(z, n / n2_sum, bounds_error=False, fill_value=0.0, kind='cubic')
         return n2
 
     def integrand_w(self, x, n, chi, chi_int, cosmo):
```
```suggestion
    @staticmethod
    def integrand_w(x, n, chi, chi_int, cosmo):
```

yymao commented on descqa/shear_test.py:
```diff
@@ -100,42 +101,41 @@ def galaxy_W(self, z, n, chi_int, cosmo, chi_recomb):
         W = np.array(val_array) * prefactor * (u.Mpc)  # now unitless
         return W
 
     def integrand_lensing_limber(self, chi, l, galaxy_W_int, chi_int, p):
```
```suggestion
    @staticmethod
    def integrand_lensing_limber(chi, l, galaxy_W_int, chi_int, p):
```

patricialarsen commented:
@yymao thanks for the suggestions! I've committed them all, and have run the test again to double check everything is still working.
